Build: make it possibly to specify subdirs in build.info

levitte opened the pull request:
This adds a keyword SUBDIRS for build.info, to be used like this:

    SUBDIRS=foo bar

This tells Configure that it should look for 'build.info' in the
relative subdirectories 'foo' and 'bar' as well.

levitte committed:
Build: make it possibly to specify subdirs in build.info

This adds a keyword SUBDIRS for build.info, to be used like this:

    SUBDIRS=foo bar

This tells Configure that it should look for 'build.info' in the
relative subdirectories 'foo' and 'bar' as well.

t-j-h commented on crypto/build.info:
```diff
@@ -1,3 +1,11 @@
+SUBDIRS=objects buffer bio stack lhash rand evp asn1 pem x509 x509v3 conf \
+        txt_db pkcs7 pkcs12 ui kdf store \
+{- my @sdirs = qw(md2 md4 md5 sha mdc2 hmac ripemd whrlpool poly1305 blake2
```
I don't think it makes sense to push this sort of perl usage in here - you should handle knocking out directories in the main perl code - not down at this level IMHO.

levitte replied:
That means we need to list these directories in two places, here and in Configure. That increases the burden of keeping things in sync.

t-j-h replied:
No it doesn't - you already filter out the list in Configure itself - you don't build a smaller list. 
You just have to treat the list coming from the build.info files in the same manner. It builds a full list from the build.info files and then you filter it out.
It just means reorganising the logic to handle that.
These should be in the build.info file - but the -no-X processing should remain in the main perl code.
The objective should be to have absolutely no actual perl hacks in the build.info files IMHO.

levitte replied:
So basically, you want something like this early in the `build.info` lookup loop:

``` perl
        continue if scalar @curd == 2 && $curd[0] eq 'crypto' && $disabled{$curd[1]};
```

and then `crypto/build.info` would have this:

```
SUBDIRS=objects buffer bio stack lhash rand evp asn1 pem x509 x509v3 conf \
        txt_db pkcs7 pkcs12 ui kdf store \
        md2 md4 md5 sha mdc2 hmac ripemd whrlpool poly1305 blake2 \
        siphash sm3 des aes rc2 rc4 rc5 idea aria bf cast camellia \
        seed sm4 chacha modes bn ec rsa dsa dh sm2 dso engine \
        err comp ocsp cms ts srp cmac ct async
```

Do I get that right?

An issue that I see with this is that someone looking at `crypto/build.info` won't see the connection between disabling certain keywords and that certain directories aren't looked into at all, and will have to dig deep in `Configure` to find out (this is the case with the current `sdirs` list as well).  That's confusing for those who do not know the inner workings of Configure.

richsalz replied:
A comment in the build.info pointing to Configure might help.  Very few do understand the inner workings of Configure these days.

levitte replied:
Ok, I'm listening...  I recalled that there is more information processing in Configure than just the filtering itself.

So yeah, I'll make the suggested changes.

richsalz commented:
I am not disagree'ing with Tim's comment, but the concept of putting SUBDIR in the build.info is a great idea.

levitte committed:
Add SUBDIRS settings in relevant build.info files

levitte committed:
Collapse different classes of macro databases

We have $config{openssl_algorithm_defines}, $config{openssl_other_defines}
and $config{openssl_thread_defines}.  These are treated exactly the same
in include/openssl/opensslconf.h.in, so having them separated into three
different databases isn't necessary, the reason for the separation being
long gone.  Therefore, we collapse them into one and the same,
$config{openssl_feature_defines}.

levitte committed:
Simplify the processing of skipped source directories

We kept a number of arrays of directory names to keep track of exactly
which directories to look for build.info.  Some of these had the extra
function to hold the directories to actually build.

With the added SUBDIRS keyword, these arrays are no longer needed.
The logic for skipping certain directories needs to be kept, though.
That is now very much simplified, and is made opportunistic.

levitte committed:
fixup! Simplify the processing of skipped source directories

levitte commented:
Approval noted, and CIs look happy.  I'll leave a grace period 'til sometime tomorrow, or perhaps monday

levitte commented:
Merged.  Thanks for the reviews!

75d47db49d41176d1f9a363f80e5a45e834563b8 Simplify the processing of skipped source directories
e0bf7c0181dbf17323dbd38dfd485970150c5244 Collapse different classes of macro databases
9654924f587bd9cd72046607f54a76c679161d26 Add SUBDIRS settings in relevant build.info files
7f73eafe2f5014ce1f915702c19ee7274e6b8c2d Build: make it possibly to specify subdirs in build.info
